### PR TITLE
stats: integrate real symbol table into stats system

### PR DESCRIPTION
**Pull request by jmarantz**
*Description*:  Switches the default from using fake symbol tables to using real symbol tables.
*Risk Level*: medium -- every effort was made to eliminate risk of contention by symbolizing stat names during initialization phases. However, some code paths may have evaded this. Using real symbol tables does appear to be significantly faster. For example, a microbenchmark for the HTTP response code stats shows approximately 40% speed improvement, in addition to a significant reduction in per-cluster memory usage.
```
----------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations
----------------------------------------------------------------------
BM_AddResponsesFakeSymtab         9116 ns         9116 ns        66687
BM_AddResponsesRealSymtab         5613 ns         5613 ns       125048
BM_ResponseTimingFakeSymtab        678 ns          678 ns      1032669
BM_ResponseTimingRealSymtab        379 ns          379 ns      1840949
``` 
*Testing*: //test/...
*Docs Changes*: should update stats.md to reflect the default change.
*Release Notes*: will need those still.
Fixes: https://github.com/envoyproxy/envoy/issues/3585, https://github.com/envoyproxy/envoy/issues/4196

### CODE REVIEW

**Comment by jmarantz** [comment]
@ambuc FYI

**Comment by jmarantz** [comment]
Got some encouraging perf results to motivate this large complicated scary PR :)

Memory usage is cut by 50% in HeapStatsThreadLocalStoreTest.Memory*:
```
--- a/test/common/stats/thread_local_store_test.cc
+++ b/test/common/stats/thread_local_store_test.cc
@@ -668,7 +668,7 @@ TEST_F(HeapStatsThreadLocalStoreTest, MemoryWithoutTls) {
-  EXPECT_LT(end_mem - start_mem, 28 * million); // actual value: 27203216 as of Oct 29, 2018
+  EXPECT_LT(end_mem - start_mem, 13 * million); // actual value: 12443472 as of Nov 7, 2018
 
 TEST_F(HeapStatsThreadLocalStoreTest, MemoryWithTls) {
@@ -691,7 +691,7 @@ TEST_F(HeapStatsThreadLocalStoreTest, MemoryWithTls) {
-  EXPECT_LT(end_mem - start_mem, 31 * million); // actual value: 30482576 as of Oct 29, 2018
+  EXPECT_LT(end_mem - start_mem, 16 * million); // actual value: 15722832 as of Nov 7, 2018
```

Speed tests also look like they doubled as well...but will confirm after a couple of corroborating runs.

**Comment by jmarantz** [comment]
Microbenchmark shows significant improvement. This differs from the siege results in that there will be no contention in the single-threaded benchmark, and of course just the pure stats lookups and does not include all the other work needed to proxy a few K of lorem ipsum.

old:
```
--------------------------------------------------------------
Benchmark                       Time           CPU Iterations
--------------------------------------------------------------
BM_StatsNoTls_mean       17363665 ns   17363797 ns         31
BM_StatsNoTls_median     17291555 ns   17291754 ns         31
BM_StatsNoTls_stddev       362025 ns     362036 ns         31
BM_StatsWithTls_mean     20593810 ns   20593997 ns         27
BM_StatsWithTls_median   20169259 ns   20169508 ns         27
BM_StatsWithTls_stddev    1412912 ns    1412939 ns         27
```

new
```
--------------------------------------------------------------
Benchmark                       Time           CPU Iterations
--------------------------------------------------------------
BM_StatsNoTls_mean        7010833 ns    7010901 ns         78
BM_StatsNoTls_median      6977452 ns    6977538 ns         78
BM_StatsNoTls_stddev       133710 ns     133713 ns         78
BM_StatsWithTls_mean      9030426 ns    9030522 ns         64
BM_StatsWithTls_median    8972650 ns    8972531 ns         64
BM_StatsWithTls_stddev     171496 ns     171491 ns         64
```

This is more than a 2x improvement in stats overhead (assuming we are looking up stats from symbolized names rather than from strings).

**Comment by jmarantz** [comment]
OK the size of this has gotten out of control and clearly will need to be broken up and re-worked in a couple of areas.  But this solves the contention issue as far as I can detect with envoy-perf/siege/siege.py, through a combination of earlier allocation of stats that were previously looked up by name per-request, and by using reader-locks and atomic ref-counts in the SymbolTable to avoid contention when looking up a name that's already in the table.

**Comment by jmarantz** [comment]
closing PR temporarily to avoid paying for CI when I don't need it.

**Comment by jmarantz** [comment]
re-opening so I can see updates to diffs.

**Comment by jmarantz** [comment]
I merged in a WiP PR (https://github.com/envoyproxy/envoy/pull/7882) to add a command-line switch to choose at startup whether to use real or fake symbol tables, as I thought that would be a good hedge against contention not caught in the other PRs that lead up to this one.

**Comment by akonradi** [comment]
What's next on this PR?

**Comment by jmarantz** [comment]
I am waiting for feedback from a couple of companies that have volunteered to turn it on and monitor perf (us being one of them).

In one case, the `/stats/recentlookups` endpoint shined a brighter light on some suboptimal patterns of allocating stats in requests in custom filters, and those are being cleaned up now.

**Comment by jmarantz** [comment]
I think this is finally ready for review (assuming CI passes).

**Comment by jmarantz** [comment]
tsan tests failures due to timeouts, one of which is eds_speed_test_benchmark_test.  On my workstation, with tsan:

 //test/common/upstream:eds_speed_test_benchmark_test                     PASSED in 156.6s

So I'm trying to re-run that; I guess maybe I've trigged that already but it will run after 'format'?

**Comment by mattklein123** [comment]
@jmarantz sorry can you merge master again to fix the conflict? The TSAN flake should be fixed.

/wait

**Comment by jmarantz** [comment]
I think I have another TSAN timeout issue as well, which I am hoping will be mitigated by https://github.com/envoyproxy/envoy/pull/11768

but let's see what happens after the merge.

**Comment by jmarantz** [comment]
@mattklein123 this looks like it passes now; I had to de-flake the http2 integration test with a wait-for-counter-ge, and increase the test-size for ads_integration_test. I suspect that tsan overhead disproportionately slows down the tests with real symbol table due to all the locks being taken, regardless of whether they contend.

File: docs/root/version_history/current.rst [hunk 1]
```diff
@@ -107,6 +107,9 @@ New Features
 * metrics service: added added :ref:`API version <envoy_v3_api_field_config.metrics.v3.MetricsServiceConfig.transport_api_version>` to explicitly set the version of gRPC service endpoint and message to be used.
 * network filters: added a :ref:`postgres proxy filter <config_network_filters_postgres_proxy>`.
 * network filters: added a :ref:`rocketmq proxy filter <config_network_filters_rocketmq_proxy>`.
+* performance: stats symbol table implementation (enabled by default; to disable it, add
+  `--use-fake-symbol-table 1` to the command-line arguments when starting Envoy).
+* prometheus stats: fix the sort order of output lines to comply with the standard.
```
Comment (mattklein123) [inline docs/root/version_history/current.rst:112]: Merge issue?

**Comment by jmarantz** [comment]
Thanks for all the help pushing that through @mattklein123! This has to be close a the record for longest PR issue->merge time delta.

**Pull request merged**